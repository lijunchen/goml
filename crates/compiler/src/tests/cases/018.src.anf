fn impl_Arith_int_add(self/0: int, other/1: int) -> int {
  int_add(self/0, other/1)
}

fn impl_Arith_int_less(self/2: int, other/3: int) -> bool {
  int_less(self/2, other/3)
}

fn impl_ToString_int_to_string(self/4: int) -> string {
  int_to_string(self/4)
}

fn impl_ToString_bool_to_string(self/5: bool) -> string {
  bool_to_string(self/5)
}

fn impl_Output_int_output(self/6: int) -> unit {
  let t4 = impl_ToString_int_to_string(self/6) in
  string_println(t4)
}

fn impl_Output_bool_output(self/7: bool) -> unit {
  let t5 = impl_ToString_bool_to_string(self/7) in
  string_println(t5)
}

fn main() -> unit {
  let a/9 = id__T_int(1) in
  let b/10 = id__T_int(2) in
  let c/11 = impl_Arith_int_add(a/9, b/10) in
  let mtmp0 = impl_Output_int_output(c/11) in
  let a/12 = id__T_int(3) in
  let b/13 = id__T_int(4) in
  let c/14 = impl_Arith_int_less(a/12, b/13) in
  let mtmp1 = impl_Output_bool_output(c/14) in
  let mtmp2 = id__T_string("abc") in
  let mtmp3 = id__T_bool(true) in
  ()
}

fn id__T_int(x/8: int) -> int {
  x/8
}

fn id__T_string(x/8: string) -> string {
  x/8
}

fn id__T_bool(x/8: bool) -> bool {
  x/8
}