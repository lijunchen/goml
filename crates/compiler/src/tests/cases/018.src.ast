trait Arith {
  fn add(Self, Self) -> int;
  fn less(Self, Self) -> bool;
}

impl Arith for int {
  fn add(self: int, other: int) -> int {
      self + other
  }
  fn less(self: int, other: int) -> bool {
      int_less(self, other)
  }
}

trait ToString {
  fn to_string(Self) -> string;
}

impl ToString for int {
  fn to_string(self: int) -> string {
      int_to_string(self)
  }
}

impl ToString for bool {
  fn to_string(self: bool) -> string {
      bool_to_string(self)
  }
}

trait Output {
  fn output(Self) -> unit;
}

impl Output for int {
  fn output(self: int) -> unit {
      string_println(to_string(self))
  }
}

impl Output for bool {
  fn output(self: bool) -> unit {
      string_println(to_string(self))
  }
}

fn id(x: T) -> T {
    x
}

fn main() {
    let a = id(1) in
    let b = id(2) in
    let c = add(a, b) in
    let _ = output(c) in
    let a = id(3) in
    let b = id(4) in
    let c = less(a, b) in
    let _ = output(c) in
    let _ = id("abc") in
    let _ = id(true) in
    ()
}

