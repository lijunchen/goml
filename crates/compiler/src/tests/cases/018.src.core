fn impl_Arith_int_add(self/0: int, other/1: int) -> int {
  int_add(self/0, other/1)
}

fn impl_Arith_int_less(self/2: int, other/3: int) -> bool {
  int_less(self/2, other/3)
}

fn impl_ToString_int_to_string(self/4: int) -> string {
  int_to_string(self/4)
}

fn impl_ToString_bool_to_string(self/5: bool) -> string {
  bool_to_string(self/5)
}

fn impl_Output_int_output(self/6: int) -> unit {
  string_println(impl_ToString_int_to_string(self/6))
}

fn impl_Output_bool_output(self/7: bool) -> unit {
  string_println(impl_ToString_bool_to_string(self/7))
}

fn id(x/8: T) -> T {
  x/8
}

fn main() -> unit {
  let a/9 = id(1) in
  let b/10 = id(2) in
  let c/11 = impl_Arith_int_add(a/9, b/10) in
  let mtmp0 = impl_Output_int_output(c/11) in
  let a/12 = id(3) in
  let b/13 = id(4) in
  let c/14 = impl_Arith_int_less(a/12, b/13) in
  let mtmp1 = impl_Output_bool_output(c/14) in
  let mtmp2 = id("abc") in
  let mtmp3 = id(true) in
  ()
}