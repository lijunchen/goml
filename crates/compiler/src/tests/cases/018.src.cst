FILE@0..1069
  TRAIT@0..81
    TraitKeyword@0..5 "trait"
    Whitespace@5..6 " "
    Uident@6..11 "Arith"
    Whitespace@11..12 " "
    LBrace@12..13 "{"
    Whitespace@13..18 "\n    "
    TRAIT_METHOD_SIG_LIST@18..81
      TRAIT_METHOD_SIG@18..43
        FnKeyword@18..20 "fn"
        Whitespace@20..21 " "
        Lident@21..24 "add"
        TYPE_LIST@24..37
          LParen@24..25 "("
          TYPE_TAPP@25..29
            Uident@25..29 "Self"
          Comma@29..30 ","
          Whitespace@30..31 " "
          TYPE_TAPP@31..35
            Uident@31..35 "Self"
          RParen@35..36 ")"
          Whitespace@36..37 " "
        Arrow@37..39 "->"
        Whitespace@39..40 " "
        TYPE_INT@40..43
          IntKeyword@40..43 "int"
      Semi@43..44 ";"
      Whitespace@44..49 "\n    "
      TRAIT_METHOD_SIG@49..76
        FnKeyword@49..51 "fn"
        Whitespace@51..52 " "
        Lident@52..56 "less"
        TYPE_LIST@56..69
          LParen@56..57 "("
          TYPE_TAPP@57..61
            Uident@57..61 "Self"
          Comma@61..62 ","
          Whitespace@62..63 " "
          TYPE_TAPP@63..67
            Uident@63..67 "Self"
          RParen@67..68 ")"
          Whitespace@68..69 " "
        Arrow@69..71 "->"
        Whitespace@71..72 " "
        TYPE_BOOL@72..76
          BoolKeyword@72..76 "bool"
      Semi@76..77 ";"
      Whitespace@77..78 "\n"
      RBrace@78..79 "}"
      Whitespace@79..81 "\n\n"
  IMPL@81..257
    ImplKeyword@81..85 "impl"
    Whitespace@85..86 " "
    Uident@86..91 "Arith"
    Whitespace@91..92 " "
    ForKeyword@92..95 "for"
    Whitespace@95..96 " "
    TYPE_INT@96..100
      IntKeyword@96..99 "int"
      Whitespace@99..100 " "
    LBrace@100..101 "{"
    Whitespace@101..106 "\n    "
    FN@106..177
      FnKeyword@106..108 "fn"
      Whitespace@108..109 " "
      Lident@109..112 "add"
      PARAM_LIST@112..136
        LParen@112..113 "("
        PARAM@113..124
          Lident@113..117 "self"
          Colon@117..118 ":"
          Whitespace@118..119 " "
          TYPE_INT@119..122
            IntKeyword@119..122 "int"
          Comma@122..123 ","
          Whitespace@123..124 " "
        PARAM@124..134
          Lident@124..129 "other"
          Colon@129..130 ":"
          Whitespace@130..131 " "
          TYPE_INT@131..134
            IntKeyword@131..134 "int"
        RParen@134..135 ")"
        Whitespace@135..136 " "
      Arrow@136..138 "->"
      Whitespace@138..139 " "
      TYPE_INT@139..143
        IntKeyword@139..142 "int"
        Whitespace@142..143 " "
      BLOCK@143..177
        LBrace@143..144 "{"
        Whitespace@144..153 "\n        "
        EXPR_BINARY@153..170
          EXPR_LIDENT@153..158
            Lident@153..157 "self"
            Whitespace@157..158 " "
          Plus@158..159 "+"
          Whitespace@159..160 " "
          EXPR_LIDENT@160..170
            Lident@160..165 "other"
            Whitespace@165..170 "\n    "
        RBrace@170..171 "}"
        Whitespace@171..177 "\n\n    "
    FN@177..254
      FnKeyword@177..179 "fn"
      Whitespace@179..180 " "
      Lident@180..184 "less"
      PARAM_LIST@184..208
        LParen@184..185 "("
        PARAM@185..196
          Lident@185..189 "self"
          Colon@189..190 ":"
          Whitespace@190..191 " "
          TYPE_INT@191..194
            IntKeyword@191..194 "int"
          Comma@194..195 ","
          Whitespace@195..196 " "
        PARAM@196..206
          Lident@196..201 "other"
          Colon@201..202 ":"
          Whitespace@202..203 " "
          TYPE_INT@203..206
            IntKeyword@203..206 "int"
        RParen@206..207 ")"
        Whitespace@207..208 " "
      Arrow@208..210 "->"
      Whitespace@210..211 " "
      TYPE_BOOL@211..216
        BoolKeyword@211..215 "bool"
        Whitespace@215..216 " "
      BLOCK@216..254
        LBrace@216..217 "{"
        Whitespace@217..226 "\n        "
        EXPR_CALL@226..252
          EXPR_LIDENT@226..234
            Lident@226..234 "int_less"
          ARG_LIST@234..252
            LParen@234..235 "("
            ARG@235..241
              EXPR_LIDENT@235..239
                Lident@235..239 "self"
              Comma@239..240 ","
              Whitespace@240..241 " "
            ARG@241..246
              EXPR_LIDENT@241..246
                Lident@241..246 "other"
            RParen@246..247 ")"
            Whitespace@247..252 "\n    "
        RBrace@252..253 "}"
        Whitespace@253..254 "\n"
    RBrace@254..255 "}"
    Whitespace@255..257 "\n\n"
  TRAIT@257..311
    TraitKeyword@257..262 "trait"
    Whitespace@262..263 " "
    Uident@263..271 "ToString"
    Whitespace@271..272 " "
    LBrace@272..273 "{"
    Whitespace@273..278 "\n    "
    TRAIT_METHOD_SIG_LIST@278..311
      TRAIT_METHOD_SIG@278..306
        FnKeyword@278..280 "fn"
        Whitespace@280..281 " "
        Lident@281..290 "to_string"
        TYPE_LIST@290..297
          LParen@290..291 "("
          TYPE_TAPP@291..295
            Uident@291..295 "Self"
          RParen@295..296 ")"
          Whitespace@296..297 " "
        Arrow@297..299 "->"
        Whitespace@299..300 " "
        TYPE_STRING@300..306
          StringKeyword@300..306 "string"
      Semi@306..307 ";"
      Whitespace@307..308 "\n"
      RBrace@308..309 "}"
      Whitespace@309..311 "\n\n"
  IMPL@311..412
    ImplKeyword@311..315 "impl"
    Whitespace@315..316 " "
    Uident@316..324 "ToString"
    Whitespace@324..325 " "
    ForKeyword@325..328 "for"
    Whitespace@328..329 " "
    TYPE_INT@329..333
      IntKeyword@329..332 "int"
      Whitespace@332..333 " "
    LBrace@333..334 "{"
    Whitespace@334..339 "\n    "
    FN@339..409
      FnKeyword@339..341 "fn"
      Whitespace@341..342 " "
      Lident@342..351 "to_string"
      PARAM_LIST@351..363
        LParen@351..352 "("
        PARAM@352..361
          Lident@352..356 "self"
          Colon@356..357 ":"
          Whitespace@357..358 " "
          TYPE_INT@358..361
            IntKeyword@358..361 "int"
        RParen@361..362 ")"
        Whitespace@362..363 " "
      Arrow@363..365 "->"
      Whitespace@365..366 " "
      TYPE_STRING@366..373
        StringKeyword@366..372 "string"
        Whitespace@372..373 " "
      BLOCK@373..409
        LBrace@373..374 "{"
        Whitespace@374..383 "\n        "
        EXPR_CALL@383..407
          EXPR_LIDENT@383..396
            Lident@383..396 "int_to_string"
          ARG_LIST@396..407
            LParen@396..397 "("
            ARG@397..401
              EXPR_LIDENT@397..401
                Lident@397..401 "self"
            RParen@401..402 ")"
            Whitespace@402..407 "\n    "
        RBrace@407..408 "}"
        Whitespace@408..409 "\n"
    RBrace@409..410 "}"
    Whitespace@410..412 "\n\n"
  IMPL@412..516
    ImplKeyword@412..416 "impl"
    Whitespace@416..417 " "
    Uident@417..425 "ToString"
    Whitespace@425..426 " "
    ForKeyword@426..429 "for"
    Whitespace@429..430 " "
    TYPE_BOOL@430..435
      BoolKeyword@430..434 "bool"
      Whitespace@434..435 " "
    LBrace@435..436 "{"
    Whitespace@436..441 "\n    "
    FN@441..513
      FnKeyword@441..443 "fn"
      Whitespace@443..444 " "
      Lident@444..453 "to_string"
      PARAM_LIST@453..466
        LParen@453..454 "("
        PARAM@454..464
          Lident@454..458 "self"
          Colon@458..459 ":"
          Whitespace@459..460 " "
          TYPE_BOOL@460..464
            BoolKeyword@460..464 "bool"
        RParen@464..465 ")"
        Whitespace@465..466 " "
      Arrow@466..468 "->"
      Whitespace@468..469 " "
      TYPE_STRING@469..476
        StringKeyword@469..475 "string"
        Whitespace@475..476 " "
      BLOCK@476..513
        LBrace@476..477 "{"
        Whitespace@477..486 "\n        "
        EXPR_CALL@486..511
          EXPR_LIDENT@486..500
            Lident@486..500 "bool_to_string"
          ARG_LIST@500..511
            LParen@500..501 "("
            ARG@501..505
              EXPR_LIDENT@501..505
                Lident@501..505 "self"
            RParen@505..506 ")"
            Whitespace@506..511 "\n    "
        RBrace@511..512 "}"
        Whitespace@512..513 "\n"
    RBrace@513..514 "}"
    Whitespace@514..516 "\n\n"
  TRAIT@516..563
    TraitKeyword@516..521 "trait"
    Whitespace@521..522 " "
    Uident@522..528 "Output"
    Whitespace@528..529 " "
    LBrace@529..530 "{"
    Whitespace@530..535 "\n    "
    TRAIT_METHOD_SIG_LIST@535..563
      TRAIT_METHOD_SIG@535..558
        FnKeyword@535..537 "fn"
        Whitespace@537..538 " "
        Lident@538..544 "output"
        TYPE_LIST@544..551
          LParen@544..545 "("
          TYPE_TAPP@545..549
            Uident@545..549 "Self"
          RParen@549..550 ")"
          Whitespace@550..551 " "
        Arrow@551..553 "->"
        Whitespace@553..554 " "
        TYPE_UNIT@554..558
          UnitKeyword@554..558 "unit"
      Semi@558..559 ";"
      Whitespace@559..560 "\n"
      RBrace@560..561 "}"
      Whitespace@561..563 "\n\n"
  IMPL@563..669
    ImplKeyword@563..567 "impl"
    Whitespace@567..568 " "
    Uident@568..574 "Output"
    Whitespace@574..575 " "
    ForKeyword@575..578 "for"
    Whitespace@578..579 " "
    TYPE_INT@579..583
      IntKeyword@579..582 "int"
      Whitespace@582..583 " "
    LBrace@583..584 "{"
    Whitespace@584..589 "\n    "
    FN@589..666
      FnKeyword@589..591 "fn"
      Whitespace@591..592 " "
      Lident@592..598 "output"
      PARAM_LIST@598..610
        LParen@598..599 "("
        PARAM@599..608
          Lident@599..603 "self"
          Colon@603..604 ":"
          Whitespace@604..605 " "
          TYPE_INT@605..608
            IntKeyword@605..608 "int"
        RParen@608..609 ")"
        Whitespace@609..610 " "
      Arrow@610..612 "->"
      Whitespace@612..613 " "
      TYPE_UNIT@613..618
        UnitKeyword@613..617 "unit"
        Whitespace@617..618 " "
      BLOCK@618..666
        LBrace@618..619 "{"
        Whitespace@619..628 "\n        "
        EXPR_CALL@628..664
          EXPR_LIDENT@628..642
            Lident@628..642 "string_println"
          ARG_LIST@642..664
            LParen@642..643 "("
            ARG@643..658
              EXPR_CALL@643..658
                EXPR_LIDENT@643..652
                  Lident@643..652 "to_string"
                ARG_LIST@652..658
                  LParen@652..653 "("
                  ARG@653..657
                    EXPR_LIDENT@653..657
                      Lident@653..657 "self"
                  RParen@657..658 ")"
            RParen@658..659 ")"
            Whitespace@659..664 "\n    "
        RBrace@664..665 "}"
        Whitespace@665..666 "\n"
    RBrace@666..667 "}"
    Whitespace@667..669 "\n\n"
  IMPL@669..777
    ImplKeyword@669..673 "impl"
    Whitespace@673..674 " "
    Uident@674..680 "Output"
    Whitespace@680..681 " "
    ForKeyword@681..684 "for"
    Whitespace@684..685 " "
    TYPE_BOOL@685..690
      BoolKeyword@685..689 "bool"
      Whitespace@689..690 " "
    LBrace@690..691 "{"
    Whitespace@691..696 "\n    "
    FN@696..774
      FnKeyword@696..698 "fn"
      Whitespace@698..699 " "
      Lident@699..705 "output"
      PARAM_LIST@705..718
        LParen@705..706 "("
        PARAM@706..716
          Lident@706..710 "self"
          Colon@710..711 ":"
          Whitespace@711..712 " "
          TYPE_BOOL@712..716
            BoolKeyword@712..716 "bool"
        RParen@716..717 ")"
        Whitespace@717..718 " "
      Arrow@718..720 "->"
      Whitespace@720..721 " "
      TYPE_UNIT@721..726
        UnitKeyword@721..725 "unit"
        Whitespace@725..726 " "
      BLOCK@726..774
        LBrace@726..727 "{"
        Whitespace@727..736 "\n        "
        EXPR_CALL@736..772
          EXPR_LIDENT@736..750
            Lident@736..750 "string_println"
          ARG_LIST@750..772
            LParen@750..751 "("
            ARG@751..766
              EXPR_CALL@751..766
                EXPR_LIDENT@751..760
                  Lident@751..760 "to_string"
                ARG_LIST@760..766
                  LParen@760..761 "("
                  ARG@761..765
                    EXPR_LIDENT@761..765
                      Lident@761..765 "self"
                  RParen@765..766 ")"
            RParen@766..767 ")"
            Whitespace@767..772 "\n    "
        RBrace@772..773 "}"
        Whitespace@773..774 "\n"
    RBrace@774..775 "}"
    Whitespace@775..777 "\n\n"
  FN@777..808
    FnKeyword@777..779 "fn"
    Whitespace@779..780 " "
    Lident@780..782 "id"
    GENERIC_LIST@782..785
      LBracket@782..783 "["
      GENERIC@783..784
        Uident@783..784 "T"
      RBracket@784..785 "]"
    PARAM_LIST@785..792
      LParen@785..786 "("
      PARAM@786..790
        Lident@786..787 "x"
        Colon@787..788 ":"
        Whitespace@788..789 " "
        TYPE_TAPP@789..790
          Uident@789..790 "T"
      RParen@790..791 ")"
      Whitespace@791..792 " "
    Arrow@792..794 "->"
    Whitespace@794..795 " "
    TYPE_TAPP@795..797
      Uident@795..796 "T"
      Whitespace@796..797 " "
    BLOCK@797..808
      LBrace@797..798 "{"
      Whitespace@798..803 "\n    "
      EXPR_LIDENT@803..805
        Lident@803..804 "x"
        Whitespace@804..805 "\n"
      RBrace@805..806 "}"
      Whitespace@806..808 "\n\n"
  FN@808..1069
    FnKeyword@808..810 "fn"
    Whitespace@810..811 " "
    Lident@811..815 "main"
    PARAM_LIST@815..818
      LParen@815..816 "("
      RParen@816..817 ")"
      Whitespace@817..818 " "
    BLOCK@818..1069
      LBrace@818..819 "{"
      Whitespace@819..824 "\n    "
      EXPR_LET@824..1068
        LetKeyword@824..827 "let"
        Whitespace@827..828 " "
        PATTERN_VARIABLE@828..830
          Lident@828..829 "a"
          Whitespace@829..830 " "
        Eq@830..831 "="
        Whitespace@831..832 " "
        EXPR_LET_VALUE@832..838
          EXPR_CALL@832..838
            EXPR_LIDENT@832..834
              Lident@832..834 "id"
            ARG_LIST@834..838
              LParen@834..835 "("
              ARG@835..836
                EXPR_INT@835..836
                  Int@835..836 "1"
              RParen@836..837 ")"
              Whitespace@837..838 " "
        InKeyword@838..840 "in"
        Whitespace@840..845 "\n    "
        EXPR_LET_BODY@845..1068
          EXPR_LET@845..1068
            LetKeyword@845..848 "let"
            Whitespace@848..849 " "
            PATTERN_VARIABLE@849..851
              Lident@849..850 "b"
              Whitespace@850..851 " "
            Eq@851..852 "="
            Whitespace@852..853 " "
            EXPR_LET_VALUE@853..859
              EXPR_CALL@853..859
                EXPR_LIDENT@853..855
                  Lident@853..855 "id"
                ARG_LIST@855..859
                  LParen@855..856 "("
                  ARG@856..857
                    EXPR_INT@856..857
                      Int@856..857 "2"
                  RParen@857..858 ")"
                  Whitespace@858..859 " "
            InKeyword@859..861 "in"
            Whitespace@861..866 "\n    "
            EXPR_LET_BODY@866..1068
              EXPR_LET@866..1068
                LetKeyword@866..869 "let"
                Whitespace@869..870 " "
                PATTERN_VARIABLE@870..872
                  Lident@870..871 "c"
                  Whitespace@871..872 " "
                Eq@872..873 "="
                Whitespace@873..874 " "
                EXPR_LET_VALUE@874..884
                  EXPR_CALL@874..884
                    EXPR_LIDENT@874..877
                      Lident@874..877 "add"
                    ARG_LIST@877..884
                      LParen@877..878 "("
                      ARG@878..881
                        EXPR_LIDENT@878..879
                          Lident@878..879 "a"
                        Comma@879..880 ","
                        Whitespace@880..881 " "
                      ARG@881..882
                        EXPR_LIDENT@881..882
                          Lident@881..882 "b"
                      RParen@882..883 ")"
                      Whitespace@883..884 " "
                InKeyword@884..886 "in"
                Whitespace@886..891 "\n    "
                EXPR_LET_BODY@891..1068
                  EXPR_LET@891..1068
                    LetKeyword@891..894 "let"
                    Whitespace@894..895 " "
                    PATTERN_WILDCARD@895..897
                      WildcardKeyword@895..896 "_"
                      Whitespace@896..897 " "
                    Eq@897..898 "="
                    Whitespace@898..899 " "
                    EXPR_LET_VALUE@899..909
                      EXPR_CALL@899..909
                        EXPR_LIDENT@899..905
                          Lident@899..905 "output"
                        ARG_LIST@905..909
                          LParen@905..906 "("
                          ARG@906..907
                            EXPR_LIDENT@906..907
                              Lident@906..907 "c"
                          RParen@907..908 ")"
                          Whitespace@908..909 " "
                    InKeyword@909..911 "in"
                    Whitespace@911..917 "\n\n    "
                    EXPR_LET_BODY@917..1068
                      EXPR_LET@917..1068
                        LetKeyword@917..920 "let"
                        Whitespace@920..921 " "
                        PATTERN_VARIABLE@921..923
                          Lident@921..922 "a"
                          Whitespace@922..923 " "
                        Eq@923..924 "="
                        Whitespace@924..925 " "
                        EXPR_LET_VALUE@925..931
                          EXPR_CALL@925..931
                            EXPR_LIDENT@925..927
                              Lident@925..927 "id"
                            ARG_LIST@927..931
                              LParen@927..928 "("
                              ARG@928..929
                                EXPR_INT@928..929
                                  Int@928..929 "3"
                              RParen@929..930 ")"
                              Whitespace@930..931 " "
                        InKeyword@931..933 "in"
                        Whitespace@933..938 "\n    "
                        EXPR_LET_BODY@938..1068
                          EXPR_LET@938..1068
                            LetKeyword@938..941 "let"
                            Whitespace@941..942 " "
                            PATTERN_VARIABLE@942..944
                              Lident@942..943 "b"
                              Whitespace@943..944 " "
                            Eq@944..945 "="
                            Whitespace@945..946 " "
                            EXPR_LET_VALUE@946..952
                              EXPR_CALL@946..952
                                EXPR_LIDENT@946..948
                                  Lident@946..948 "id"
                                ARG_LIST@948..952
                                  LParen@948..949 "("
                                  ARG@949..950
                                    EXPR_INT@949..950
                                      Int@949..950 "4"
                                  RParen@950..951 ")"
                                  Whitespace@951..952 " "
                            InKeyword@952..954 "in"
                            Whitespace@954..959 "\n    "
                            EXPR_LET_BODY@959..1068
                              EXPR_LET@959..1068
                                LetKeyword@959..962 "let"
                                Whitespace@962..963 " "
                                PATTERN_VARIABLE@963..965
                                  Lident@963..964 "c"
                                  Whitespace@964..965 " "
                                Eq@965..966 "="
                                Whitespace@966..967 " "
                                EXPR_LET_VALUE@967..978
                                  EXPR_CALL@967..978
                                    EXPR_LIDENT@967..971
                                      Lident@967..971 "less"
                                    ARG_LIST@971..978
                                      LParen@971..972 "("
                                      ARG@972..975
                                        EXPR_LIDENT@972..973
                                          Lident@972..973 "a"
                                        Comma@973..974 ","
                                        Whitespace@974..975 " "
                                      ARG@975..976
                                        EXPR_LIDENT@975..976
                                          Lident@975..976 "b"
                                      RParen@976..977 ")"
                                      Whitespace@977..978 " "
                                InKeyword@978..980 "in"
                                Whitespace@980..985 "\n    "
                                EXPR_LET_BODY@985..1068
                                  EXPR_LET@985..1068
                                    LetKeyword@985..988 "let"
                                    Whitespace@988..989 " "
                                    PATTERN_WILDCARD@989..991
                                      WildcardKeyword@989..990 "_"
                                      Whitespace@990..991 " "
                                    Eq@991..992 "="
                                    Whitespace@992..993 " "
                                    EXPR_LET_VALUE@993..1003
                                      EXPR_CALL@993..1003
                                        EXPR_LIDENT@993..999
                                          Lident@993..999 "output"
                                        ARG_LIST@999..1003
                                          LParen@999..1000 "("
                                          ARG@1000..1001
                                            EXPR_LIDENT@1000..1001
                                              Lident@1000..1001 "c"
                                          RParen@1001..1002 ")"
                                          Whitespace@1002..1003 " "
                                    InKeyword@1003..1005 "in"
                                    Whitespace@1005..1011 "\n\n    "
                                    EXPR_LET_BODY@1011..1068
                                      EXPR_LET@1011..1068
                                        LetKeyword@1011..1014 "let"
                                        Whitespace@1014..1015 " "
                                        PATTERN_WILDCARD@1015..1017
                                          WildcardKeyword@1015..1016 "_"
                                          Whitespace@1016..1017 " "
                                        Eq@1017..1018 "="
                                        Whitespace@1018..1019 " "
                                        EXPR_LET_VALUE@1019..1029
                                          EXPR_CALL@1019..1029
                                            EXPR_LIDENT@1019..1021
                                              Lident@1019..1021 "id"
                                            ARG_LIST@1021..1029
                                              LParen@1021..1022 "("
                                              ARG@1022..1027
                                                EXPR_STR@1022..1027
                                                  Str@1022..1027 "\"abc\""
                                              RParen@1027..1028 ")"
                                              Whitespace@1028..1029 " "
                                        InKeyword@1029..1031 "in"
                                        Whitespace@1031..1036 "\n    "
                                        EXPR_LET_BODY@1036..1068
                                          EXPR_LET@1036..1068
                                            LetKeyword@1036..1039 "let"
                                            Whitespace@1039..1040 " "
                                            PATTERN_WILDCARD@1040..1042
                                              WildcardKeyword@1040..1041 "_"
                                              Whitespace@1041..1042 " "
                                            Eq@1042..1043 "="
                                            Whitespace@1043..1044 " "
                                            EXPR_LET_VALUE@1044..1053
                                              EXPR_CALL@1044..1053
                                                EXPR_LIDENT@1044..1046
                                                  Lident@1044..1046 "id"
                                                ARG_LIST@1046..1053
                                                  LParen@1046..1047 "("
                                                  ARG@1047..1051
                                                    EXPR_BOOL@1047..1051
                                                      TrueKeyword@1047..1051 "true"
                                                  RParen@1051..1052 ")"
                                                  Whitespace@1052..1053 " "
                                            InKeyword@1053..1055 "in"
                                            Whitespace@1055..1065 "\n    \n    "
                                            EXPR_LET_BODY@1065..1068
                                              EXPR_UNIT@1065..1068
                                                LParen@1065..1066 "("
                                                RParen@1066..1067 ")"
                                                Whitespace@1067..1068 "\n"
      RBrace@1068..1069 "}"