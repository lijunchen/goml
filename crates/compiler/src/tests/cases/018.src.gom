package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    panic("missing: " + s)
    return struct{}{}
}

func impl_Arith_int_add(self__0 int, other__1 int) int {
    var ret6 int
    ret6 = int_add(self__0, other__1)
    return ret6
}

func impl_Arith_int_less(self__2 int, other__3 int) bool {
    var ret7 bool
    ret7 = int_less(self__2, other__3)
    return ret7
}

func impl_ToString_int_to_string(self__4 int) string {
    var ret8 string
    ret8 = int_to_string(self__4)
    return ret8
}

func impl_ToString_bool_to_string(self__5 bool) string {
    var ret9 string
    ret9 = bool_to_string(self__5)
    return ret9
}

func impl_Output_int_output(self__6 int) struct{} {
    var ret10 struct{}
    var t4 string = impl_ToString_int_to_string(self__6)
    ret10 = string_println(t4)
    return ret10
}

func impl_Output_bool_output(self__7 bool) struct{} {
    var ret11 struct{}
    var t5 string = impl_ToString_bool_to_string(self__7)
    ret11 = string_println(t5)
    return ret11
}

func main0() struct{} {
    var ret12 struct{}
    var a__9 int = id__T_int(1)
    var b__10 int = id__T_int(2)
    var c__11 int = impl_Arith_int_add(a__9, b__10)
    impl_Output_int_output(c__11)
    var a__12 int = id__T_int(3)
    var b__13 int = id__T_int(4)
    var c__14 bool = impl_Arith_int_less(a__12, b__13)
    impl_Output_bool_output(c__14)
    id__T_string("abc")
    id__T_bool(true)
    ret12 = struct{}{}
    return ret12
}

func id__T_int(x__8 int) int {
    var ret13 int
    ret13 = x__8
    return ret13
}

func id__T_string(x__8 string) string {
    var ret14 string
    ret14 = x__8
    return ret14
}

func id__T_bool(x__8 bool) bool {
    var ret15 bool
    ret15 = x__8
    return ret15
}

func main() {
    main0()
}
