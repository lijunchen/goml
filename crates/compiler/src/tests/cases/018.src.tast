impl Arith for int{
  fn add(self/0: int, other/1: int) -> int {
    (self/0 : int) + (other/1 : int)
  }
  fn less(self/2: int, other/3: int) -> bool {
    int_less((self/2 : int), (other/3 : int))
  }
}

impl ToString for int{
  fn to_string(self/4: int) -> string {
    int_to_string((self/4 : int))
  }
}

impl ToString for bool{
  fn to_string(self/5: bool) -> string {
    bool_to_string((self/5 : bool))
  }
}

impl Output for int{
  fn output(self/6: int) -> unit {
    string_println(to_string((self/6 : int)))
  }
}

impl Output for bool{
  fn output(self/7: bool) -> unit {
    string_println(to_string((self/7 : bool)))
  }
}

fn id(x/8: T) -> T {
  (x/8 : T)
}

fn main() -> unit {
  let a/9: int = id(1) in
  let b/10: int = id(2) in
  let c/11: int = add((a/9 : int), (b/10 : int)) in
  let _ : unit = output((c/11 : int)) in
  let a/12: int = id(3) in
  let b/13: int = id(4) in
  let c/14: bool = less((a/12 : int), (b/13 : int)) in
  let _ : unit = output((c/14 : bool)) in
  let _ : string = id("abc") in
  let _ : bool = id(true) in
  ()
}