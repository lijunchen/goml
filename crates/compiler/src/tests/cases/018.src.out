()
====
3
true

====
3
true
